### PR TITLE
WIP: fix maxfev & TODO, save 1 fct eval

### PR DESCRIPTION
**Pull request by dtnguyen2**
# Note

This PR limits the number of function evaluation in the  ```gridsearch``` optimization method if the default option ```maxfev``` is set to an integer (assuming it is > 0, but does not check).
 
Not sure why this stupid method was written since the number of function of evaluation is num^npar where num is the option parameter ```num``` and npar is the number of free parameters.

### CODE REVIEW

**Comment by dtnguyen2** [comment]
Here is a reason why I think shipping the ```gridsearch``` method is a bad idea.  Take the simple _basic-fit.py_ script that is shipped with sherpa (but comment out the second fit cause the sherpa command ```m2.offset = 4.3979 * m2.c1```  will exceed the limits)
```Python
(testing-PR) [dtn@devel12 sherpatest]$ git diff basic-fit.py
diff --git a/sherpatest/basic-fit.py b/sherpatest/basic-fit.py
index d39ef07..ca003a0 100644
--- a/sherpatest/basic-fit.py
+++ b/sherpatest/basic-fit.py
@@ -6,11 +6,12 @@ set_model(polynom1d.m1)
 thaw(m1.c1)
 thaw(m1.c2)
 thaw(m1.c3)
-
+set_method('gridsearch')
+print(get_model())
 fit()
 
-set_model(polynom1d.m2)
-thaw(m2.c1)
-m2.offset = 4.3979 * m2.c1
+# set_model(polynom1d.m2)
+# thaw(m2.c1)
+# m2.offset = 4.3979 * m2.c1
 
-fit()
+# fit()
```

So after 65536 function evaluations searching on a hyper-grid of size 16 within the limits [-3.4e38, 3.4e38], on gets:

```Python
(testing-PR) [dtn@devel12 sherpatest]$ python basic-fit.py 
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: Failed to import any usable sherpa.plotting backend. Plotting routines will not be available.
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
polynom1d.m1
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   m1.c0        thawed            1 -3.40282e+38  3.40282e+38           
   m1.c1        thawed            0 -3.40282e+38  3.40282e+38           
   m1.c2        thawed            0 -3.40282e+38  3.40282e+38           
   m1.c3        thawed            0 -3.40282e+38  3.40282e+38           
   m1.c4        frozen            0 -3.40282e+38  3.40282e+38           
   m1.c5        frozen            0 -3.40282e+38  3.40282e+38           
   m1.c6        frozen            0 -3.40282e+38  3.40282e+38           
   m1.c7        frozen            0 -3.40282e+38  3.40282e+38           
   m1.c8        frozen            0 -3.40282e+38  3.40282e+38           
   m1.offset    frozen            0 -3.40282e+38  3.40282e+38           
Dataset               = 1
Method                = gridsearch
Statistic             = chi2
Initial fit statistic = 20919.3
Final fit statistic   = 1.64264e+41 at function evaluation 65536
Data points           = 11
Degrees of freedom    = 7
Probability [Q-value] = 0
Reduced statistic     = 2.34663e+40
Change in statistic   = 1.64264e+41
   m1.c0          -3.40282e+38
   m1.c1          -3.40282e+38
   m1.c2          2.4954e+38  
   m1.c3          -2.26855e+37
WARNING: parameter value m1.c0 is at its minimum boundary -3.4028234663852886e+38
WARNING: parameter value m1.c1 is at its minimum boundary -3.4028234663852886e+38
```

BTW,  quite a few of the ```XSpec``` models have the limit set at the floating point max value (~3.4e38).

```bash
(testing-PR) [dtn@devel12 xspec]$ pwd
/export/sherpa/sherpa/astro/xspec
(testing-PR) [dtn@devel12 xspec]$ grep hugeval __init__.py | wc
    148    1313   11194
```

The minimum using ```LevMar```:
```Python
Dataset               = 1
Method                = levmar
Statistic             = chi2
Initial fit statistic = 20919.3
Final fit statistic   = 30.8491 at function evaluation 10
Data points           = 11
Degrees of freedom    = 7
Probability [Q-value] = 6.62869e-05
Reduced statistic     = 4.40701
Change in statistic   = 20888.4
   m1.c0          1.49843      +/- 0.0520846   
   m1.c1          0.1447       +/- 0.0413773   
   m1.c2          0.0322936    +/- 0.00874997  
   m1.c3          -0.00277729  +/- 0.000523425 
```

```bash
(testing-PR) [dtn@devel12 xspec]$ pwd
/export/sherpa/sherpa/astro/xspec
(testing-PR) [dtn@devel12 xspec]$ grep hugeval __init__.py | wc
    148    1313   11194
```
The free parameters are near unity, so my first thought was: hey why not restrict the limits to within a short range from the starting free parameters?  for example:0

```Python
diff --git a/sherpa/optmethods/optfcts.py b/sherpa/optmethods/optfcts.py
index be18e14a..86ee6c61 100644
--- a/sherpa/optmethods/optfcts.py
+++ b/sherpa/optmethods/optfcts.py
@@ -338,7 +338,7 @@ def difevo_nm(fcn, x0, xmin, xmax, ftol, maxfev, verbose, seed,
 
 
 def grid_search(fcn, x0, xmin, xmax, num=16, sequence=None, numcores=1,
-                maxfev=None, ftol=EPSILON, method=None, verbose=0):
+                maxfev=None, ftol=EPSILON, method=None, verbose=0, factor=2):
     """Grid Search optimization method.
 
     This method evaluates the fit statistic for each point in the
@@ -420,6 +420,17 @@ def grid_search(fcn, x0, xmin, xmax, num=16, sequence=None, numcores=1,
 
     if sequence is None:
         ranges = []
+        myxmin = []
+        myxmax = []
+        for lo, val, hi in zip(xmin, x, xmax):
+            if val == 0:
+                myxmin.append(max(lo, val - factor))
+        print('myxmin =', myxmin)
+        print('myxmax =', myxmax)
         for index in range(npar):
             ranges.append([xmin[index], xmax[index]])
         if method is not None:
```
Even though the limits are now within a much smaller range
```Python
myxmin = [-1.0, -2.0, -2.0, -2.0]
myxmax = [3.0, 2.0, 2.0, 2.0]
```
One is still no where near the optimum solution:

```Python
(testing-PR) [dtn@devel12 sherpatest]$ python basic-fit.py 
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: Failed to import any usable sherpa.plotting backend. Plotting routines will not be available.
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
myxmin = [-1.0, -2.0, -2.0, -2.0]
myxmax = [3.0, 2.0, 2.0, 2.0]
Dataset               = 1
Method                = gridsearch
Statistic             = chi2
Initial fit statistic = 20919.3
Final fit statistic   = 20919.3 at function evaluation 65537
Data points           = 11
Degrees of freedom    = 7
Probability [Q-value] = 0
Reduced statistic     = 2988.47
Change in statistic   = 0
   m1.c0          1           
   m1.c1          0           
   m1.c2          0           
   m1.c3          0 
```

Never mind for now that the ```gridsearch``` is inefficient.  As you know, finding the minimum  in 1-d is a trivial problem.  However, finding a minimum in multi-dim is not.  The ```gridsearch``` basically tries to reduce a minimization problem in high dimensions by reducing it to a one dimension which is not wise.

File: sherpa/optmethods/optfcts.py [hunk 1]
```diff
@@ -404,6 +404,8 @@ def make_sequence(ranges, N):
 
         grid = numpy.mgrid[list_ranges]
         mynfev = pow(N, npar)
+        if maxfev is not None:
+            mynfev = min(maxfev, mynfev)
         grid = list(map(numpy.ravel, grid))
```
[the remark this answers is not in the document]
Comment (harlantc) [reply sherpa/optmethods/optfcts.py:409]: I gave this a look out of curiosity. A quick guess is that codecov is not comparing the lines it says it is.
https://app.codecov.io/gh/sherpa/sherpa/compare/1433/tree/sherpa/optmethods/optfcts.py
You'll see that the diff is way off, about 300 lines have been shifted down on the new diff... I can't say for sure why that's happening, but if I was to hazard guess, it looks like codecov's diff started comparing some of the commented out code in the docstrings with code elsewhere, and that threw it off.

This is super annoying and if it keeps happening I'll see if there's a way to improve the diff codecov uses.

[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/optmethods/optfcts.py:409]: Yuch. Ta for looking @harlantc

**Comment by dtnguyen2** [comment]
I was told that the optimization method called ```gridsearch``` was written for template fitting.  So this PR should be moved to an issue (not sure how to do this right now but will try to figure it out).

**Comment by hamogu** [comment]
See #1478

**Pull request closed without merging**